x/feegrant/periodic_fee.go

```diff
@@ -49,21 +49,27 @@ func (a *PeriodicAllowance) Accept(ctx sdk.Context, fee sdk.Coins, _ []sdk.Msg)
 
 // tryResetPeriod will check if the PeriodReset has been hit. If not, it is a no-op.
 // If we hit the reset period, it will top up the PeriodCanSpend amount to
-// min(PeriodicSpendLimit, a.Basic.SpendLimit) so it is never more than the maximum allowed.
+// min(PeriodSpendLimit, Basic.SpendLimit) so it is never more than the maximum allowed.
 // It will also update the PeriodReset. If we are within one Period, it will update from the
 // last PeriodReset (eg. if you always do one tx per day, it will always reset the same time)
 // If we are more then one period out (eg. no activity in a week), reset is one Period from the execution of this method
 func (a *PeriodicAllowance) tryResetPeriod(blockTime time.Time) {
 	if blockTime.Before(a.PeriodReset) {
 		return
 	}
-	// set CanSpend to the lesser of PeriodSpendLimit and the TotalLimit
+
+	// set PeriodCanSpend to the lesser of Basic.SpendLimit and PeriodSpendLimit
 	if _, isNeg := a.Basic.SpendLimit.SafeSub(a.PeriodSpendLimit); isNeg {
 		a.PeriodCanSpend = a.Basic.SpendLimit
 	} else {
 		a.PeriodCanSpend = a.PeriodSpendLimit
 	}
 
+	// if Basic.SpendLimit is not set, reset to PeriodSpendLimit
+	if a.Basic.SpendLimit.Empty() {
+		a.PeriodCanSpend = a.PeriodSpendLimit
+	}
```

Comment:
I can add this check to the condition above but I thought it might be nice to separate for clarity.

Reply:
Isn't it a logic error to put here?

Say I create a grant of 200token total, with a max periodic spend limit of 100token/mo, and the grantee uses everything.
- Expected a beginning of 1st month: `a.Basic.SpendLimit == 200, a.PeriodCanSpend == 100`
- Expected at end of 1st month: `a.Basic.SpendLimit == 100, a.PeriodCanSpend == 0`
- Expected a beginning of 2nd month: `a.Basic.SpendLimit == 100, a.PeriodCanSpend == 100`
- Expected at end of 2nd months: `a.Basic.SpendLimit == 0, a.PeriodCanSpend == 0`
- But with this code, after two months: `a.Basic.SpendLimit == 0, a.PeriodCanSpend == 100`

Reply:
:facepalm: Yes it is. I've added the empty check to the condition above.

Reply:
At the end of the second month, the grant would be removed. I will take another look and double check.

Reply:
Just to note, we're not actually updating the value of `a.Basic.SpendLimit` within the `tryResetPeriod` method; we are doing that within the `Accept` method.

If `a.Basic.SpendLimit` is set and the result of `a.Basic.SpendLimit.SafeSub(a.PeriodSpendLimit)` is negative, we will set `a.PeriodCanSpend` to `a.Basic.SpendLimit`.

If `a.Basic.SpendLimit` is not set, the result of `a.Basic.SpendLimit.SafeSub(a.PeriodSpendLimit)` is always negative, and `a.PeriodCanSpend` is always set to empty.

I manually tested the above scenario when I was writing the code and double checked again and the grant is removed after the second month (2 minutes in my test).

Reply:
Just to note, we're not actually updating the value of `a.Basic.SpendLimit` within the `tryResetPeriod` method; we are doing that within the `Accept` method. Maybe this is causing some confusion.

If `a.Basic.SpendLimit` is set and the result of `a.Basic.SpendLimit.SafeSub(a.PeriodSpendLimit)` is negative, we will set `a.PeriodCanSpend` to `a.Basic.SpendLimit`.

If `a.Basic.SpendLimit` is not set, the result of `a.Basic.SpendLimit.SafeSub(a.PeriodSpendLimit)` is always negative, and `a.PeriodCanSpend` is always set to empty.

I believe this is covered in the tests. I also manually tested the above scenario when I was writing the code and double checked again and the grant is removed after the second month (2 minutes in my test).

Reply:
I consolidated this into the condition that was already there. The same logic applies so let me know if you think I might still be overlooking something.